[Jed] Step 3-3: 옵저버 패턴 적용, 리팩토링

junu0516 opened the pull request:
## 작업 내용
- [x] ViewController-Plane 간 NotificationCenter을 사용한 옵저버 패턴 적용 
- [x] Plane에서의 ViewController 참조 제거
- [x] Plane 내부 불필요한 속성, 서브스크립트 제거
    
## 학습키워드
- NotificationCenter
- Observer Pattern
    
## 고민과 해결
### 2단계 리팩토링
- Plane 내부의 불필요한 속성 및 서브스크립트를 제거했습니다.
- __`RectangleViewFactory`__ 를 추가하여 ViewController에서 개별 RectangleView를 만들지 않고 팩토리에서 리턴하는 객체를 서브 뷰로 추가하도록 수정했습니다.
 
### 옵저버 패턴 적용 
- 기존에는 StylerView를 통한 사용자 입력이 들어오면 사용자입력-ViewController-CanvasView-ViewController-StylerView-ViewControllerPlane의 복잡한 과정을 거쳐 순차적으로 입출력이 일어나도록 했지만, __`Observer Pattern`__ 을 적용하여 __`사용자 입력-ViewController-Plane-CanvasView&StylerView`__ 으로 과정을 단순화했습니다.
    - 기존에는 StylerView, CanvasView에서 각각 출력이 완료되면 delegate method를 통해 ViewController에 출력이 끝났음을 알린 후, ViewController가 후속동작으로 Plane의 메소드를 다시 호출하는 식이었지만 __`NotificationCenter`__ 을 사용하여 굳이 delegate method 호출 없이 모델의 값이 변화하면 알아서 ViewController의 메소드를 호출하는 식으로 변경했습니다. 
    - 또한, Plane에서 __`PlaneDelegate`__ 를 통해 모델에서의 값 변경이 끝났음을 ViewController 알리지 않고 미리 등록해둔 옵저버에 알리도록 수정했기에 더 이상 필요 없어진 PlaneDelegate 프로토콜을 없애고 __Plane에서 ViewController을 참조하던 것을 제거했습니다.__
    
### 질문
- 현재 아래와 같은 방식으로 사용자가 CanvasView 상의 어느 한 지점을 터치하면 위와 같이 제스처를 인식하여 사각형 모델 및 뷰를 처리하도록 하고 있는데, 지난 번 피드백에서 이에 대해 __`self.currentlyTouchedView`__ 와 같이 굳이 뷰컨트롤러에 속성을 두지 말고 __`Plane 에서 선택된 Rectangle 모델과 매칭되는 것을 처리`__ 하는 것이 더 낫다고 해주신 부분에 대해 질문이 있습니다.

  ```swift
      func gestureRecognizer(_ gestureRecognizer: UIGestureRecognizer, shouldReceive touch: UITouch) -> Bool {
          /*
             1. 우선 터치된 곳에 해당하는 touch.view의 타입이 RectangleView인 경우에는,
                현재 터치된 사각형뷰를 임시로 저장한 currentlyTouchedView의 인스턴스로 할당
             2. 이후 Plane에서 터치된 x,y를 가지고 유효한 모델이 내부에 존재하는 지 검증되면 
                 이후 뷰에서 뷰컨트롤러로 값 변경 요청이 들어오면 currentlyTouchedView를 사용하여 사각형 뷰의 배경색 혹은 투명도를 바꿀 수 있도록 함
             4. 단, 이후 터치된 곳에 해당하는 touch.view가 RectangleView가 아닌 경우에는, 
                 사각형이 없는 부분을 터치한 것이므로 currentlyTouchedView를 다시 nil로 초기화
           */
          
          if let touchedView = touch.view as? RectangleView{
              self.currentlyTouchedView = touchedView
          }else{
              self.currentlyTouchedView = nil
          }
          let touchedPoint = touch.location(in: self.canvasView)
          self.plane.findMatchingRectangleModel(x: touchedPoint.x, y: touchedPoint.y)
          return true
      }
  ```
- 위에서 __`currentlyTouchedView`__ 의 경우에는 __`touch.view`__ 가 사실상 __`gestureRecognizer`__ 메소드 내에서만 유효한 인스턴스이기 때문에, 만일 Plane에서 유효한 Rectangle 모델을 찾은 후, 화면을 수정하기 위한 뷰컨트롤러 메소드를 호출했을 때 이를 사용할 수 없어 뷰컨트롤러의 속성값으로 임시로 가지고 있도록 했습니다.
- 지난번 피드백에 대해 제가 이해한 바로는 이런 식으로 속성에 입력값을 두지 않는 것이 좋기 때문에 이를 제거하고 다른 방법을 적용해야 하는 것으로 우선 알고 있는데, 이에 대해  __`touch.view`__ 를 활용하는 것 외에는 뷰에서 터치한 사각형 뷰와 Plane에서 찾은 사각형 모델을 매칭하는 방법이 아직까지 안떠오르는데 혹시 이에 대해 조금 더 힌트를 얻을 수 있을까요?

junu0516 committed:
Refactor: ViewController, CanvasView, StylerView 메소드 파라미터 타입 변경

- 사각형 배경색 랜덤으로 바꿀 때, 파라미터로 넘기는 타입을 [Double]에서 UIColor로 변경
- [Double]타입 인스턴스는 Plane을 호출하는 순간에만 생성해서 넘기도록 함

junu0516 committed:
Add: RectangleViewFactory.swift 추가

junu0516 committed:
Refactor: RectangleViewFactory에서 리턴하는 RectangleView 사용하도록 수정

- UIView를 ViewController에서 직접 생성하지 않고 별도의 팩토리에서 생성해서 받아쓰도록 수정
- RectangleViewFactory의 메소드에서 RectangleView 타입 인스턴스 리턴
- 리턴한 해당 인스턴스를 ViewController에서 CanvasView로 넘겨서 화면에 출력

junu0516 committed:
Refactor: Plane 내부 불필요한 속성, 서브스크립트 제거

- rectangleIndex 배열 제거
- 정수타입 index 파라미터로 모델 찾는 서브스크립트 제거

junu0516 committed:
Refactor: PlaneDelegate 참조 제거(Observer Pattern 적용)

- ViewController에 NotificationCenter 속성 추가 및 옵저버 등록
- 기존의 Plane 내부의 PlaneDelegate 참조 제거
- 모델 변경 이벤트 발생하면 NotificationCenter.post 호출

junu0516 committed:
Refactor: CanvasView, StylerView 배경색, 투명도 바꾸는 로직 수정(Observer Pattern 적용)

- NotificationCenter에 사각형 배경색, 투명도 바꾸는 옵저버 추가
- 사용자 액션 입력(뷰) -> 뷰컨트롤러 -> 모델 -> 뷰 컨트롤러 -> 뷰 출력 의 흐름으로 모두 수정
- 기존의 StylerViewDelegate, CanvasViewDelegate의 출력 끝나고 호출하던 ***Completed() 메소드 모두 제거
- 제거한 해당 메소드를 모두 NotificationCenter.post로 대체

junu0516 committed:
Refactor: CanvasViewDelegate, StylerViewDelegate 안쓰이는 메소드 제거

기존의 StylerViewDelegate, CanvasViewDelegate의 출력 끝나고 호출하던 ***Completed() 메소드 모두 제거

junu0516 committed:
Refactor: Rectangle 배경색 변경시  파라미터 타입을 Rectangle.Color로 수정

- ViewController에서 Plane으로 파라미터 넘겨줄 때 [Double]이 아닌 Rectangle.Color로 넘기도록 수정
- ViewController에서 View로 출력 위한 파라미터 넘겨줄 때는 Rectangle.Color 값으로 UIColor 인스턴스 생성해서 넘겨주도록 수정

junu0516 committed:
Chore: 질문할 부분 주석 추가

godrm commented:
> 위에서 currentlyTouchedView 의 경우에는 touch.view 가 사실상 gestureRecognizer 메소드 내에서만 유효한 인스턴스이기 때문에, 만일 Plane에서 유효한 Rectangle 모델을 찾은 후, 화면을 수정하기 위한 뷰컨트롤러 메소드를 호출했을 때 이를 사용할 수 없어 뷰컨트롤러의 속성값으로 임시로 가지고 있도록 했습니다.

네 입력에서 뷰를 찾을 수 있고, 그 뷰가 선택됐더라도 속성으로 남겨놓기 보다는 모델에서 변화한 것을 찾아서 매칭해주는 게 유연한 구조입니다. 
하지만 매칭하는 게 꼭 뷰 컨트롤러에서만 해야 하는가 이 부분도 고민스럽기는 합니다. 
그럼에도 불구하고 두 가지 정보를 모두 접근할 수 있는 게 뷰 컨트롤러라서 그렇습니다.

godrm commented on DrawingApp/DrawingApp/Controller/ViewController.swift:
```diff
@@ -8,22 +8,28 @@ class ViewController: UIViewController{
     private var stylerView: StylerView?
     private var plane: Plane = Plane()
     private var currentlyTouchedView: RectangleView?
+    private var notificationCenter = NotificationCenter.default
```
싱글톤을 사용할 때 이렇게 참조를 만드는 게 꼭 좋은 건 아닙니다. 
NotificationCenter는 그렇지 않지만 default 라는게 항상 같은 인스턴스가 아니고 바뀌는 경우도 있거든요.

godrm commented on DrawingApp/DrawingApp/Controller/ViewController.swift:
```diff
@@ -8,22 +8,28 @@ class ViewController: UIViewController{
     }
+
+    
+    private func initializeNotificationCenter(){
+        self.notificationCenter.addObserver(self, selector: #selector(rectangleFoundFromPlane(_:)), name: .rectangleFoundFromPlane, object: nil)
+        self.notificationCenter.addObserver(self, selector: #selector(rectangleNotFoundFromPlane), name: .rectangleNotFoundFromPlane, object: nil)
+        self.notificationCenter.addObserver(self, selector: #selector(addingRectangleCompleted(_:)), name: .rectangleAdded, object: nil)
+        self.notificationCenter.addObserver(self, selector: #selector(updateSelectedRecntalgeViewColor(_:)), name: .rectangleColorUpdated , object: nil)
+        self.notificationCenter.addObserver(self, selector: #selector(updateSelectedRectangleViewAlpha(_:)), name: .rectangleAlphaUpdated, object: nil)
```
기존 델리게이트로 이벤트를 받던 것과 이렇게 NotificationCenter를 통한 방법이 어떤 차이가 있는 것 같은가요?

junu0516 replied:
제가 우선 학습한 바로는 Plane 내부에서 ViewController을 참조하지 않기 때문에 결합도가 더 느슨해지는 점이 있었습니다! 
서로 더 느슨하게 연결되다 보니 하나를 수정할 때 다른 하나를 수정하는 부분을 염두에 둬야 하는 정도가 더 적어져서 좋은 것 같습니다.
(델리게이트와 옵저버의 차이는 계속 학습중입니다...)

godrm commented on DrawingApp/DrawingApp/Controller/ViewController.swift:
```diff
@@ -57,54 +63,81 @@ class ViewController: UIViewController{
     }
     
+    @objc func addingRectangleCompleted(_ notification: Notification) {
+        guard let rectangle = notification.object as? Rectangle else { return }
```
어.. NotificationCenter에서 object는 sender의 의미입니다. 학습 영상에 잘 나와있을 겁니다.
만약 데이터를 보내고 싶으면 object가 아니라 userInfo에 넣어야 합니다. 이 부분은 개선해보세요

godrm commented on DrawingApp/DrawingApp/Controller/ViewController.swift:
```diff
@@ -8,22 +8,28 @@ class ViewController: UIViewController{
+
+    
+    private func initializeNotificationCenter(){
+        self.notificationCenter.addObserver(self, selector: #selector(rectangleFoundFromPlane(_:)), name: .rectangleFoundFromPlane, object: nil)
+        self.notificationCenter.addObserver(self, selector: #selector(rectangleNotFoundFromPlane), name: .rectangleNotFoundFromPlane, object: nil)
```
NotificationCenter에서 object는 sender의 의미라서 누가 보내는 것인지 지정이 필요하다면 object를 활용하셔야 합니다.

godrm commented on DrawingApp/DrawingApp/Controller/ViewController.swift:
```diff
@@ -57,54 +63,81 @@ class ViewController: UIViewController{
+    }
+    
+    @objc func updateSelectedRecntalgeViewColor(_ notification: Notification){
+        guard let newColor = notification.object as? Rectangle.Color else { return }
```
NotificationCenter에서 object는 sender의 의미입니다. 만약 데이터를 보내고 싶으면 object가 아니라 userInfo에 넣어야 합니다. 이 부분은 개선해보세요

godrm commented on DrawingApp/DrawingApp/Controller/ViewController.swift:
```diff
@@ -57,54 +63,81 @@ class ViewController: UIViewController{
+        guard let canvasView = self.canvasView else { return }
+        
+        let newUIColor = UIColor(red: newColor.r, green: newColor.g, blue: newColor.b, alpha: 1)
+        let newHexString = "#\(String(Int(newColor.r*255), radix: 16))\(String(Int(newColor.g*255), radix: 16))\(String(Int(newColor.b*255), radix: 16))"
```
이런 표현, 형식을 맞추는 것도 여기서 할 일입니다만, 이게 잘 바뀌었는지 어떻게 실행해보지 않고 검증할 수 있을까요. 
그럴려면 이런 변환 코드도 분리하는 게 좋습니다.

junu0516 replied:
아무래도 해당 부분은 별도의 메소드로 빼서 테스트 코드로 검증할 수 있도록 해야 될 것 같습니다!

godrm commented on DrawingApp/DrawingApp/Controller/ViewController.swift:
```diff
@@ -57,54 +63,81 @@ class ViewController: UIViewController{
+           3. 단, 이후 터치된 곳에 해당하는 touch.view가 RectangleView가 아닌 경우에는, 사각형이 없는 부분을 터치한 것이므로 currentlyTouchedView를 다시 nil로 초기화
+         */
+        
         if let touchedView = touch.view as? RectangleView{
             self.currentlyTouchedView = touchedView
+        }else{
+            self.currentlyTouchedView = nil
         }
         let touchedPoint = touch.location(in: self.canvasView)
```
이 부분이군요. 입력한 뷰를 남겨 놓는 것이 편할 수 있지만, 입력과 출력을 분리해서 생각하는 게 좋습니다
이벤트는 그 뷰가 받았지만, 그걸 거기에 출력한다고 보장하지 않고 구현하는 게 좋습니다. 두 개를 같이 보는 게 오히려 복잡해집니다

junu0516 replied:
음.. 속성 없이 선택된 모델과 뷰를 매칭할 수 있는 방법을 좀 더 고민해보겠습니다

godrm commented on DrawingApp/DrawingApp/Controller/ViewController.swift:
```diff
@@ -117,63 +150,27 @@ extension ViewController: CanvasViewDelegate{
-    }
+}
+
+extension Notification.Name{
 
+    static let rectangleAdded = Notification.Name("rectangleAdded")
+    static let rectangleFoundFromPlane = Notification.Name("rectangleFoundFromPlane")
+    static let rectangleNotFoundFromPlane = Notification.Name("rectangleNotFoundFromPlane")
+    static let rectangleColorUpdated = Notification.Name("rectangleColorUpdated")
+    static let rectangleAlphaUpdated = Notification.Name("rectangleAlphaUpdated")
```
이 노티는 누가 보내는 것이고 누가 받는 것인가요
그것을 생각해보고 어느 타입에 내장에서 선언하는 게 좋을지 고민해보세요. 이렇게 모아놓는 것보다는 관련있는 타입에 모아놓는 게 응집도가 높은 겁니다

junu0516 replied:
현재는 모두 Plane 내부에서 모델값의 변경이 완료되면 보내는 노티로 했습니다. 아무래도 보내는 쪽에 선언하는 것이 해당 extension의 의도를 이해하기 더 좋을 것 같아서 이동하려 합니다!

godrm commented on DrawingApp/DrawingApp/Model/Plane.swift:
```diff
@@ -2,34 +2,47 @@ import Foundation
+        var opacity = opacity
+        guard let selectedRectangleId = self.selectedRectangleId else { return }
+        guard let rectangle = self.rectangles[selectedRectangleId] else { return }
+        if(opacity == 10){
+            opacity = opacity - 1
         }
+        rectangle.alpha = Rectangle.Alpha.allCases[opacity]
```
뭔가 범위가 일치하지 않는건가요? allCases를 쓰는 것보다 Alpha.init()을 쓰는 것도 좋았을 것 같네요. 배열의 인덱스로 접근하기 보다는 값 자체로 접근하는 게 어떨까요?

junu0516 replied:
열거형으로 선언해서 케이스가 총 10개인데, opacity = 10일 경우에는 allCases에서 인덱스값을 10으로 넣어서 배열의 범위(0~9)를 초과해서 해당 부분만 예외적으로 처리하도록 했습니다.
이 부분은 외부가 Rectangle.Alpha 내부에서 처리하도록 수정해보겠습니다!

godrm commented on DrawingApp/DrawingApp/Model/Plane.swift:
```diff
@@ -52,17 +65,9 @@ struct Plane:CustomStringConvertible{
     subscript(x: Double = 0, y: Double = 0)-> Rectangle?{
   
-        for id in rectangleIndex.reversed(){
+        for id in self.rectangles.keys{
             guard let rectangle = rectangles[id] else { continue }
             if(isRectangleInsideTheRange(x: x, y: y, rectangle: rectangle)){
                 return rectangle
```
self.rectangles가 Dictionary라서 이렇게 찾는거군요. 이런 경우는 사각형이 중첩되어 있으면 키값으로 찾다보면 위아래를 구분하지 못할 수 있지 않을까요? 예를 들어 아래 사각형이 먼저 찾아진다거나...

junu0516 replied:
self.rectangles를 배열로 하고 인덱스의 마지막부터 탐색해야 중첩될 경우 맨 위의 사각형부터 선택할 수 있을 것 같습니다!

godrm commented on DrawingApp/DrawingApp/View/RectangleViewFactory.swift:
```diff
@@ -0,0 +1,17 @@
+
+class RectangleViewFactory{
+    
+    static func createRectangleView(rectangle: Rectangle)-> RectangleView{
```
여기서 Rectangle 와 RectangleView를 구체 타입에 의존하지 않도록 개선할 수 있다면 도전해보세요.

junu0516 replied:
넵 도전해보겠습니다

godrm reviewed (changes requested):
옵저버를 잘 적용하셨는데, sender의 역할을 하는 object 부분은 개선해보세요
Plane에서는 `self.rectangles.keys`로 접근하는 부분도 개선해보세요
나머지는 다음 단계에서 해도 될 것 같습니다.

junu0516 committed:
Chore: 주석 삭제

junu0516 committed:
Refactor: Plane.rectangles 타입 변경

- rectangles 타입을 배열로 변경
- selectedRectangleId 속성을 제거하고, selectedRectangleIndex 인덱스 정수값으로 속성 대체
- 인덱스 도출하는 서브스크립트 추가

junu0516 committed:
Refactor: 싱글톤 객체 참조하는 속성 모두 제거

- NotificationCenter.default를 참조하던 notificationCenter 프로퍼티 변수 선언 모두 제거

junu0516 committed:
Refactor: Notification.Name extension 선언을 Plane 쪽으로 이동

junu0516 committed:
Refactor: Observer userInfo, object 수정

- 같은 참조값을 사용하도록 Plane을 구조체에서 클래스로 변경
- addObserver의 object를 self.plane으로 한정
- rectangle, opacity, color 등은 모두 userInfo로 넘기도록 수정

junu0516 commented:
@godrm  우선 말씀하신 옵저버의 sender과 Plane 내부에서 rectangle.id 사용하는 부분 수정했습니다! 확인 부탁드립니다.

godrm commented on DrawingApp/DrawingApp/Model/Plane.swift:
```diff
@@ -1,48 +1,61 @@
 import Foundation
 
-struct Plane:CustomStringConvertible{
+extension Notification.Name{
+
+    static let rectangleAdded = Notification.Name("rectangleAdded")
+    static let rectangleFoundFromPlane = Notification.Name("rectangleFoundFromPlane")
+    static let rectangleNotFoundFromPlane = Notification.Name("rectangleNotFoundFromPlane")
+    static let rectangleColorUpdated = Notification.Name("rectangleColorUpdated")
+    static let rectangleAlphaUpdated = Notification.Name("rectangleAlphaUpdated")
+}
```
이 노티가 Plane이 보내는 거라면 Plane.xxx 형태로 내부 namespace를 따르도록 하는게 더 깔끔합니다. 참고하세요

junu0516 replied:
넵 내부에 네임스페이스 선언하겠습니다

godrm commented on DrawingApp/DrawingApp/Model/Plane.swift:
```diff
@@ -58,17 +71,18 @@ struct Plane:CustomStringConvertible{
-    subscript(id: Rectangle.Id)-> Rectangle?{
-        if let rectangle = rectangles[id]{
-            return rectangle
+    subscript(id: Rectangle.Id)-> Int?{
+        for index in 0..<rectangles.count {
+            if(rectangles[index].id == id){
+                return index
+            }
         }
         return nil
     }
     
     subscript(x: Double = 0, y: Double = 0)-> Rectangle?{
   
```
subscript가 두 개나 있네요. subscript 는 어떤 의도에서 쓰게 될까요?  
index로 접근하는 게 아니라 plane에 Rectangle.id로 접근한다. plane에게 id가 중요한 값일가요?
아래 x, y 포인트로 접근하는 것도 있는데 그냥 메소드로 제공하는 것과 어떤 차이가 있을까요? 조금은 근본적인 질문이네요.
오히려 매개변수만 `plane[xxx]` 또는 `plane[x:, y:]` 형태가 되서 의도를 알기 더 어렵지는 않을까요?

junu0516 replied:
사실 맨 처음에는 뷰컨트롤러에서 서브스크립트를 사용하려고 정의했는데 나중에 수정하면서 뷰컨트롤러에서 메소드로 Plane에 접근하도록 수정하면서 위처럼 사용하게 되버렸습니다. 
전달값을 사각형 모델의 아이디값으로 넘기는 서브스크립트를 하나 더 추가한 이유는 이전에 rectangles 속성의 타입이 딕셔너리여서 key를 넘겨서 선택된 뷰와 매칭되는 모델을 한 번에 찾으려고 했기 때문입니다.
이 부분은 selectedRectangleId, Index... 등등의 값 없이 Plane 내부에서 범위로 뷰와 모델을 매칭하는 식으로 수정할 것이기 때문에 결과적으로 plane 내부에서 rectnagleId를 사용하지 않을 것 같으므로 결과적으로 서브스크립트는 x,y만 넘기는 것 하나만 남겨두는 식으로 수정하려 합니다!

godrm commented:
다른 부분을 계속 하려는 건가요? 아니면 이 상태로 머지하면 될까요?

junu0516 commented:
나머지 피드백 주신 부분은 다음단계 진행하면서 같이 하겠습니다! 머지 부탁드립니다.

junu0516 committed:
Refactor: Plane 내부에 별도의 NotificationName  namespace 선언

junu0516 committed:
Chore: 오타 수정
